domains/Accounts/Controllers/AccountsLoginController.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace Domains\Accounts\Controllers;
+
+use App\Http\Controllers\Controller;
+use Domains\Accounts\Models\User;
+use Illuminate\Http\Request;
+use Illuminate\Http\Response;
+use Illuminate\Support\Facades\Hash;
+
+class AccountsLoginController extends Controller
+{
+    public function __invoke(Request $request): Response
+    {
+        $this->validate($request, [
+            'email'    => ['required', 'string', 'email', 'max:255'],
+            'password' => ['required', 'string'],
+        ]);
+
+        $user = User::where('email', $request->email)->first();
+
+        if ($user && Hash::check($request->password, $user->password)) {
+            $token    = $user->createToken($user->email)->accessToken;
+            $response = ['access_token' => $token];
+            return new Response($response, Response::HTTP_OK);
+        }
+
+        return new Response(['message' => 'The authentication credentials are wrong'], Response::HTTP_UNAUTHORIZED);
```

Comment:
We should probably inverse this operation. If we take the return early approach and check if the `$user` is null or the password check fails, we won't need the interpreter to validate both operations in the logical condition, and we can return a validation error sooner.

Btw, the response status code is not correct. The action isn't unauthorized, but the validation failed. So, either a 422 or a general 400 status code could be a better fit in this case. I advise the use of the `ValidationException` and let the framework handle that correctly.

Reply:
@josepostiga,

I am sorry but I did not understand what you mean when you say to reverse the logic.
Firstly, I'm validating that the data is sent and that it's in compliance.
Then I'll validate that the user exists in the database and that his password is correct.
In the IF statement, if the user does not exist, he fails and doesn't continue with the rest of the execution.

Reply:
The return early approach is very useful for simplifying logical statements. 

Look at the if statement: we're checking for two statements to return true in order for the execution logic to go through what we call the "happy path". When I said to inverse the logic, was to switch the semantics of that if statement, in order to check if the user is null or if the password isn't correct.

PHP has an internal short circuit system for logical conditionals, and it'll automatically jump out of executing any more parts of the if statement as soon as one of those parts returns `true` (that's why we'll compare with falsy values). With that in mind, we can do `if (!$user || !Hash::check($request->password, $user->password))`, and throw a validation exception inside this new if. 

The rest of the code, where it generates and returns the token can go after the if.

Reply:
Okay, I get it. Changed!